architecture/2024-08-bicepdeployment-controller.md

```diff
@@ -0,0 +1,550 @@
+-->
+
+#### BicepDeployment Custom Resource Definition
+
+The `BicepDeployment` resource will be a new Kubernetes CRD that will be used to contain the Bicep template and parameters for deploying Radius resources. The CRD will have the following fields:
+
+```go
+// BicepDeploymentSpec defines the desired state of a BicepDeployment
+type BicepDeploymentSpec struct {
+  // Template is the Bicep template that defines the resources to deploy.
+  Template string `json:"template"`
+
+  // Parameters is the Bicep parameters file that specifies the parameters for the template.
+  Parameters string `json:"parameters"`
+
+  // Modules is a list of Bicep modules that are used in the template.
+  Modules []BicepDeploymentModule `json:"modules,omitempty"`
+}
+
+// Deploy again and check output resources. if they are the same, then we are good to go. if there are resources that are missing, then delete them and update resources 
+
+// BicepDeploymentPhrase is a string representation of the current status of a Bicep Deployment.
+type BicepDeploymentPhrase string
+
+const (
+	// BicepDeploymentPhraseUpdating indicates that the Bicep Deployment is being updated.
+	BicepDeploymentPhraseUpdating BicepDeploymentPhrase = "Updating"
+
+	// BicepDeploymentPhraseReady indicates that the Bicep Deployment is ready.
+	BicepDeploymentPhraseReady BicepDeploymentPhrase = "Ready"
+
+	// BicepDeploymentPhraseFailed indicates that the Bicep Deployment has failed.
+	BicepDeploymentPhraseFailed BicepDeploymentPhrase = "Failed"
+
+	// BicepDeploymentPhraseDeleting indicates that the Bicep Deployment is being deleted.
+	BicepDeploymentPhraseDeleting BicepDeploymentPhrase = "Deleting"
+
+	// BicepDeploymentPhraseDeleted indicates that the Bicep Deployment has been deleted.
+	BicepDeploymentPhraseDeleted BicepDeploymentPhrase = "Deleted"
+)
+
+// BicepDeploymentStatus defines the observed state of the Bicep Deployment.
+type BicepDeploymentStatus struct {
+	// ObservedGeneration is the most recent generation observed for this Bicep Deployment. It corresponds to the Deployment's generation, which is updated on mutation by the API Server.
+	ObservedGeneration int64 `json:"observedGeneration,omitempty" protobuf:"varint,1,opt,name=observedGeneration"`
+
+	// Scope is the resource ID of the scope.
+	Scope string `json:"scope,omitempty"`
+
+	// Resource is the resource ID of the deployment.
+	Resource string `json:"resource,omitempty"`
+
+	// Operation tracks the status of an in-progress provisioning operation.
+	Operation *ResourceOperation `json:"operation,omitempty"`
+
+	// Phrase indicates the current status of the Bicep Deployment.
+	Phrase BicepDeploymentPhrase `json:"phrase,omitempty"`
+
+  // Description is a human-readable description of the status of the Bicep Deployment.
+  Description string `json:"description,omitempty"`
+
+  // Resources is a list of resources that are created by the BicepDeployment.
+  Resources []BicepDeploymentResource `json:"resources,omitempty"`
+}
```

Comment:
Do we need `outputs`? to be captured here?

Reply:
Unless I'm missing it, you'll also need a means of surfacing errors.